components/addressmanager/src/lib.rs

```diff
@@ -55,124 +55,123 @@ impl AddressManager {
-                let gateway = igd::search_gateway(Default::default()).ok()?;
-                let ip = IpAddress::new(gateway.get_external_ip().ok()?);
-                if !ip.is_publicly_routable() {
-                    info!("Non-publicly routable external ip from gateway using upnp {} not added to store", ip);
-                    return None;
-                }
-                info!("Got external ip from gateway using upnp: {ip}");
+            None => {}
+        };
 
-                let default_port = self.config.default_p2p_port();
+        itertools::Either::Right(self.routable_addresses_from_net_interfaces())
+    }
 
-                let normalized_p2p_listen_address = self.config.p2p_listen_address.normalize(default_port);
-                let local_addr = if normalized_p2p_listen_address.ip.is_unspecified() {
-                    SocketAddr::new(local_ip_address::local_ip().unwrap(), normalized_p2p_listen_address.port)
-                } else {
-                    normalized_p2p_listen_address.into()
-                };
-
-                match gateway.add_port(
-                    igd::PortMappingProtocol::TCP,
-                    default_port,
-                    local_addr,
-                    UPNP_DEADLINE_SEC as u32,
-                    UPNP_REGISTRATION_NAME,
-                ) {
-                    Ok(_) => {
-                        info!("Added port mapping to default external port: {ip}:{default_port}");
-                        Some((
-                            NetAddress { ip, port: default_port },
-                            Some(ExtendHelper { gateway, local_addr, external_port: default_port }),
-                        ))
-                    }
-                    Err(AddPortError::PortInUse {}) => {
-                        let port = gateway
-                            .add_any_port(igd::PortMappingProtocol::TCP, local_addr, UPNP_DEADLINE_SEC as u32, UPNP_REGISTRATION_NAME)
-                            .ok()?;
-                        info!("Added port mapping to random external port: {ip}:{port}");
-                        Some((NetAddress { ip, port }, Some(ExtendHelper { gateway, local_addr, external_port: port })))
-                    }
-                    Err(err) => {
-                        warn!("error adding port: {err}");
-                        None
-                    }
-                }
-            }
-            None => None,
+    fn routable_addresses_from_net_interfaces(&self) -> impl Iterator<Item = NetAddress> + '_ {
+        // check whatever was passed as listen address (if routable)
+        // otherwise(listen_address === 0.0.0.0) check all interfaces
+        let listen_address = self.config.p2p_listen_address.normalize(self.config.default_p2p_port());
+        if listen_address.ip.is_publicly_routable() {
+            info!("Publicly routable local address found: {}", listen_address.ip);
+            itertools::Either::Left(Some(listen_address).into_iter())
```

Comment:
It's nicer to use `iter::once(listen_address)`

Reply:
unfortunately it will produce the error of incompatible types
`Option.into_iter()` gives me
 ```
struct Item<A> {
    opt: Option<A>,
}
```
for both None and Some. `iter::empty` and `iter::once` gives 2 different types.
to make it works I have to implement or find Either3 with predefined iterator implementation


and actually my first attempt was with `iter::` and then I changed it to Option

Reply:
in the last commit I changed it to Left(Left(...) Left(Right(..._) and Right(...). not sure which variant is better/cleaner/nicer